File: examples/basic_usage.js
// Mirrors the Rust `basic_usage.rs` example: ingest a couple of hours of readings, roll them up,
// and read aggregates back at different granularities.
//
// Run with:
//   node examples/basic_usage.js
// (after `npm run build`, so the native addon exists)

const { Engine } = require("..");

const engine = new Engine({
  dimensions: ["browser"],
  measures: [
    {
      name: "visits",
      valueType: "f64",
      aggregates: ["sum", "count", "min", "max", "average"],
    },
  ],
});

// 1. Ingest some raw samples. Every sample only ever touches a minute-level bucket.
const start = Date.UTC(2026, 5, 1, 9, 0, 0); // 2026-06-01T09:00:00Z
const readings = [
  [0, 21.5, "Firefox"],
  [1, 22.0, "Firefox"],
  [3, 19.8, "Firefox"],
  [15, 23.1, "Firefox"],
  [47, 20.4, "Firefox"],
  [61, 25.0, "Firefox"], // rolls into the next hour
  [62, 24.7, "Firefox"],
  [125, 18.9, "Firefox"], // rolls into a third hour
];

for (const [minuteOffset, value, browser] of readings) {
  engine.ingest(start + minuteOffset * 60_000, [value], [browser]);
}
console.log(
  `ingested ${readings.length} raw samples into ${engine.bucketCount("minute")} minute buckets`
);

// 2. Roll everything up. This only merges bucket states upward — it never re-reads a sample.
engine.rollup();

// 3. Read aggregates back at whatever granularity is useful.
console.log("\nPer-hour breakdown:");
for (const bucket of engine.buckets("hour")) {
  for (const group of bucket.groups) {
    const agg = group.measures[0]; // "visits" is measure index 0
    const from = new Date(bucket.startMs).toISOString().slice(11, 16);
    const to = new Date(bucket.endMs).toISOString().slice(11, 16);
    console.log(
      `  [${group.dimensionValues.join(",")} ${from} .. ${to}) ` +
        `sum=${agg.sum.toFixed(2)} count=${agg.count} ` +
        `min=${agg.min.toFixed(2)} max=${agg.max.toFixed(2)} avg=${agg.average.toFixed(2)}`
    );
  }
}

console.log("\nWhole-day total (rolled all the way up):");
const dayStart = Date.UTC(2026, 5, 1, 0, 0, 0);
const dayBucket = engine.buckets("day").find((b) => b.startMs === dayStart);
if (dayBucket) {
  const agg = dayBucket.groups[0].measures[0];
  console.log(`  count=${agg.count} sum=${agg.sum.toFixed(2)} average=${agg.average.toFixed(2)}`);
}

// 4. Ad-hoc range query for the first two hours only — merges whichever buckets already exist,
//    without storing anything new.
console.log("\nAd-hoc query for the first two hours only:");
const range = engine.queryRange("hour", start, start + 2 * 3_600_000, 0);
console.log(`  count=${range.count} sum=${range.sum.toFixed(2)}`);

// 5. Grouped query — one row per distinct "browser" value.
console.log("\nGrouped by browser (whole range):");
for (const row of engine.queryRangeGrouped("hour", start, start + 3 * 3_600_000, 0, ["browser"])) {
  console.log(`  browser=${row.dimensionValues[0]} sum=${row.aggregate.sum.toFixed(2)}`);
}

// 6. Retention: keep only the last hour of minute-level detail.
console.log("\nRetention: keeping only the last hour of minute-level detail");
const boundedEngine = new Engine({
  dimensions: ["browser"],
  measures: [{ name: "visits", valueType: "f64", aggregates: ["sum", "count"] }],
  retention: [{ level: "minute", maxAgeMs: 3_600_000 }],
});
for (const [minuteOffset, value, browser] of readings) {
  boundedEngine.ingest(start + minuteOffset * 60_000, [value], [browser]);
}
console.log(`  before prune: ${boundedEngine.bucketCount("minute")} minute buckets`);
boundedEngine.prune();
console.log(`  after prune:  ${boundedEngine.bucketCount("minute")} minute buckets`);

File: accreta/examples/basic_usage.rs
//! Basic usage: ingest a stream of samples spanning a couple of hours, roll them up through
//! every bucket level, and read back aggregates at different granularities.
//!
//! Run with:
//! ```text
//! cargo run --example basic_usage
//! ```

use accreta::aggregate_set::Schema;
use accreta::aggregates::{Average, Count, Max, Min, Sum};
use accreta::bucket::BucketLevel;
use accreta::engine::Engine;
use accreta::measures::MeasureId;
use chrono::{Duration, TimeZone, Utc};

fn main() {
    // 1. Describe which aggregates every bucket should track. Adding a new one later (e.g.
    //    Variance) is just another `.with::<T>()` call — nothing else in this example
    //    would need to change. Each aggregate is registered under its own type name
    //    (`T::NAME`), so reading it back later is done by type via `AggregateSet::get::<T>()`
    //    rather than a string key — there's no `get::<Sum>("sum")` call to typo or mismatch.
    let mut builder = Schema::builder();
    builder
        .dimension("browser")
        .measure("visits")
        .with::<Sum<f64>>()
        .with_any::<Count>()
        .with::<Min<f64>>()
        .with::<Max<f64>>()
        .with::<Average<f64>>();
    let schema = builder.build().unwrap();

    let mut engine = Engine::new(schema.clone());

    // 2. Ingest some raw samples. Every sample only ever touches a minute-level bucket.
    let start = Utc.with_ymd_and_hms(2026, 6, 1, 9, 0, 0).unwrap();
    let readings = [
        (0, 21.5, "Firefox"),
        (1, 22.0, "Firefox"),
        (3, 19.8, "Firefox"),
        (15, 23.1, "Firefox"),
        (47, 20.4, "Firefox"),
        (61, 25.0, "Firefox"), // rolls into the next hour
        (62, 24.7, "Firefox"),
        (125, 18.9, "Firefox"), // rolls into a third hour
    ];
    for (minute_offset, value, dim_val) in readings {
        _ = engine.ingest(
            start + Duration::minutes(minute_offset),
            vec![value],
            vec![dim_val],
        );
    }
    println!(
        "ingested {} raw samples into {} minute buckets",
        readings.len(),
        engine.bucket_count(BucketLevel::Minute)
    );

    // 3. Roll everything up. This never re-reads a Sample — only merges bucket states upward
    //    through Hour -> Day -> Week -> Month -> Year.
    engine.rollup();

    // 4. Read aggregates back at whatever granularity is useful.
    println!("\nPer-hour breakdown:");
    for bucket in engine.buckets(BucketLevel::Hour) {
        for group in bucket.groups() {
            let dim_key = group.0;
            let aggs = group.1;
            for agg in aggs {
                let avg = agg.get::<Average<f64>>().unwrap();
                println!(
                    "  [{:?} {} .. {}) sum={:>6.2} count={:<2} min={:>5.2} max={:>5.2} avg={:>5.2}",
                    dim_key.values(),
                    bucket.start().format("%H:%M"),
                    bucket.end().format("%H:%M"),
                    agg.get::<Sum<f64>>().unwrap().value(),
                    agg.get::<Count>().unwrap().value(),
                    agg.get::<Min<f64>>().unwrap().value().unwrap(),
                    agg.get::<Max<f64>>().unwrap().value().unwrap(),
                    avg.sum() / avg.count() as f64
                );
            }
        }
    }

    println!("\nWhole-day total (rolled all the way up):");
    let day_start = BucketLevel::Day.truncate(start);
    let day = engine.bucket(BucketLevel::Day, day_start).unwrap();
    let aggs = day.groups().next().unwrap().1;
    for agg in aggs {
        let avg = agg.get::<Average<f64>>().unwrap();
        println!(
            "  count={} sum={:.2} average={:.2}",
            agg.get::<Count>().unwrap().value(),
            agg.get::<Sum<f64>>().unwrap().value(),
            avg.sum() / avg.count() as f64,
        );
    }

    // 5. Ad-hoc range queries merge whichever buckets already exist at a given level, without
    //    storing anything new — handy for "give me the last N hours" style queries.
    println!("\nAd-hoc query for the first two hours only:");
    let range = engine
        .query_range(
            BucketLevel::Hour,
            start,
            start + Duration::hours(2),
            MeasureId(0),
        )
        .unwrap();
    println!(
        "  count={} sum={:.2}",
        range.get::<Count>().unwrap().value(),
        range.get::<Sum<f64>>().unwrap().value(),
    );

    // 6. Retention: an Engine keeps every bucket forever by default, which is fine for a bounded
    //    batch job like this one but not for long-running ingestion. `Retention` bounds memory
    //    at whichever levels you configure; `prune` is the explicit, opt-in step that actually
    //    discards old buckets (rollup itself never deletes anything).
    println!("\nRetention: keeping only the last hour of minute-level detail");
    let schema_for_retention = {
        let mut b = Schema::builder();
        b.dimension("browser")
            .measure("visits")
            .with::<Sum<f64>>()
            .with_any::<Count>();
        b.build().unwrap()
    };
    let policy = accreta::Retention::new().keep(BucketLevel::Minute, Duration::hours(1));
    let mut bounded_engine = Engine::with_retention(schema_for_retention, policy);
    for (minute_offset, value, dim_val) in readings {
        _ = bounded_engine.ingest(
            start + Duration::minutes(minute_offset),
            vec![value],
            vec![dim_val],
        );
    }
    println!(
        "  before prune: {} minute buckets",
        bounded_engine.bucket_count(BucketLevel::Minute)
    );
    bounded_engine.prune();
    println!(
        "  after prune:  {} minute buckets (older than 1h before the newest sample were dropped)",
        bounded_engine.bucket_count(BucketLevel::Minute)
    );
}
